File: expli-benchmarks/STAMP/intruder/intruder.cpp
// STAMP/intruder benchmark — explicit TM API port
// Matches the plugin intruder_bench.hpp algorithm.
//
// Uses std:: containers with mutex for serialization, matching the
// plugin's tm_serialize_lock/unlock inside TX functions.

#include "expli_tm_api/tm_api.hpp"

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <mutex>
#include <queue>
#include <random>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

#include "../../tests/benchmark_test.hpp"

using PRNG = std::mt19937_64;

struct Packet {
    long flow_id;
    int fragment_id;
    int num_fragments;
    int length;
    std::string data;
};

struct DecodedFlow {
    long flow_id;
    std::string data;
    int num_fragments_received;
    std::vector<std::string> fragments;
};

struct IntruderData {
    std::queue<Packet> packet_queue;
    std::unordered_map<long, DecodedFlow> decoder_map;
    std::queue<DecodedFlow> decoded_queue;
    std::vector<std::string> dictionary;
    int num_flows;
    int max_data_length;
    int percent_attack;
    int total_attacks;
};

static IntruderData g_data;
static std::mutex g_mutex;          // matches tm_serialize_lock

static int g_num_threads = 4;
static int g_percent_attack = 10;
static int g_max_data_length = 16;
static int g_num_flows = 1024;
static int g_seed = 1;

static void parse_args(int argc, char* argv[]) {
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "-p") && i + 1 < argc) g_num_threads = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-a") && i + 1 < argc) g_percent_attack = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-l") && i + 1 < argc) g_max_data_length = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-n") && i + 1 < argc) g_num_flows = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-s") && i + 1 < argc) g_seed = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-h")) {
            fprintf(stderr, "Usage: %s -p <threads> -a <pct_attack> -l <max_len> -n <flows> -s <seed>\n", argv[0]);
            exit(0);
        }
    }
}

// ── Attack detection (non-TX, pure string ops) ────────────
static bool detect_attack(const std::string& data, const std::vector<std::string>& dict) {
    std::string lower;
    lower.resize(data.size());
    for (size_t i = 0; i < data.size(); i++)
        lower[i] = (data[i] >= 'A' && data[i] <= 'Z') ? (data[i] - 'A' + 'a') : data[i];
    for (const auto& sig : dict) {
        if (lower.find(sig) != std::string::npos)
            return true;
    }
    return false;
}

// ── Serialized queue/decode operations ────────────────────
static Packet get_packet() {
    std::lock_guard<std::mutex> lock(g_mutex);
    if (g_data.packet_queue.empty()) return {-1, -1, -1, -1, ""};
    Packet p = g_data.packet_queue.front();
    g_data.packet_queue.pop();
    return p;
}

static void process_decoder(const Packet& pkt) {
    std::lock_guard<std::mutex> lock(g_mutex);
    auto& decoder_map = g_data.decoder_map;
    auto it = decoder_map.find(pkt.flow_id);
    if (it == decoder_map.end()) {
        DecodedFlow df;
        df.flow_id = pkt.flow_id;
        df.num_fragments_received = 0;
        df.fragments.resize(pkt.num_fragments);
        decoder_map[pkt.flow_id] = df;
        it = decoder_map.find(pkt.flow_id);
    }

    it->second.fragments[pkt.fragment_id] = pkt.data;
    it->second.num_fragments_received++;

    if (it->second.num_fragments_received == pkt.num_fragments) {
        std::string full;
        for (int i = 0; i < pkt.num_fragments; i++)
            full += it->second.fragments[i];
        it->second.data = full;
        g_data.decoded_queue.push(it->second);
        decoder_map.erase(it);
    }
}

static DecodedFlow get_complete() {
    std::lock_guard<std::mutex> lock(g_mutex);
    if (g_data.decoded_queue.empty()) return {-1, "", 0, {}};
    DecodedFlow df = g_data.decoded_queue.front();
    g_data.decoded_queue.pop();
    return df;
}

// ── Worker thread ─────────────────────────────────────────
static std::atomic<uint64_t> g_total_ops{0};

static void worker() {
    expli::TM<int>::thread_init();

    for (;;) {
        Packet pkt = get_packet();
        if (pkt.flow_id >= 0)
            process_decoder(pkt);

        DecodedFlow df = get_complete();
        if (df.flow_id >= 0) {
            bool attack = detect_attack(df.data, g_data.dictionary);
            (void)attack;
            g_total_ops.fetch_add(1, std::memory_order_relaxed);
        } else if (pkt.flow_id < 0) {
            break;
        }
    }

    expli::TM<int>::thread_exit();
}

static void test_cli_flags() {
    printf("  Testing CLI flags...\n");
    int save_p = g_num_threads, save_a = g_percent_attack, save_l = g_max_data_length;
    int save_n = g_num_flows, save_s = g_seed;
    TEST_EQ(g_num_threads, 4, "default threads");
    TEST_EQ(g_percent_attack, 10, "default attack pct");
    TEST_EQ(g_max_data_length, 16, "default max data length");
    TEST_EQ(g_num_flows, 1024, "default num flows");
    TEST_EQ(g_seed, 1, "default seed");
    const char* test_args[] = {"prog", "-p", "2", "-a", "50", "-l", "32", "-n", "8", "-s", "99"};
    parse_args(11, (char**)test_args);
    TEST_EQ(g_num_threads, 2, "override threads");
    TEST_EQ(g_percent_attack, 50, "override attack pct");
    TEST_EQ(g_max_data_length, 32, "override max data length");
    TEST_EQ(g_num_flows, 8, "override num flows");
    TEST_EQ(g_seed, 99, "override seed");
    g_num_threads = save_p; g_percent_attack = save_a; g_max_data_length = save_l;
    g_num_flows = save_n; g_seed = save_s;
    if (test_result() != 0) exit(1);
}

static void test_rng() {
    printf("  Testing RNG determinism...\n");
    test_rng_determinism<PRNG>();
    if (test_result() != 0) exit(1);
}

static void test_logic() {
    printf("  Testing intruder logic...\n");
    // Verify attack classification
    std::vector<std::string> dict = {"attack", "root", "system"};
    TEST_ASSERT(detect_attack("this is an attack string", dict), "detect attack");
    TEST_ASSERT(detect_attack("ROOT access granted", dict), "detect case-insensitive");
    TEST_ASSERT(!detect_attack("benign data here", dict), "no false positive");
    // Test packet generation determinism
    PRNG rng(42);
    int num_packets = 0, attack_count = 0;
    for (long flow = 1; flow <= 20; flow++) {
        bool is_attack = (int)(rng() % 100) < 10;
        if (is_attack) attack_count++;
        int payload_len = (int)(rng() % 16) + 1;
        int np = (int)(rng() % payload_len) + 1;
        num_packets += np;
    }
    TEST_ASSERT(num_packets > 0, "generated packets");
    if (test_result() != 0) exit(1);
}

int main(int argc, char* argv[]) {
    if (argc > 1 && strcmp(argv[1], "--test") == 0) {
        printf("Running self-tests for intruder...\n");
        test_cli_flags();
        test_rng();
        test_logic();
        printf("All tests passed.\n");
        return 0;
    }
    parse_args(argc, argv);

    expli::TM<int>::init();

    g_data.num_flows = g_num_flows;
    g_data.max_data_length = g_max_data_length;
    g_data.percent_attack = g_percent_attack;
    g_data.total_attacks = 0;

    g_data.dictionary = {"about", "attack", "back", "root", "system", "access",
                         "all", "after", "also", "and", "any", "are", "but",
                         "can", "come", "could", "did", "do", "each", "find",
                         "first", "for", "from", "get", "go", "has", "have",
                         "her", "here", "him", "his", "how", "into", "its",
                         "just", "know", "like", "look", "make", "man", "may",
                         "more", "most", "must", "new", "no", "not", "now",
                         "old", "one", "only", "other", "our", "out", "over",
                         "own", "part", "people", "said", "say", "see", "she",
                         "shell", "should", "site", "some", "such", "take",
                         "than", "that", "their", "them", "then", "there"};

    PRNG rng(g_seed);
    int total_packets = 0;

    for (long flow = 1; flow <= g_num_flows; flow++) {
        bool is_attack = (int)(rng() % 100) < g_percent_attack;
        std::string payload;

        if (is_attack) {
            int sig_idx = (int)(rng() % g_data.dictionary.size());
            payload = g_data.dictionary[sig_idx];
            g_data.total_attacks++;
        } else {
            int len = (int)(rng() % g_max_data_length) + 1;
            payload.resize(len);
            for (int i = 0; i < len; i++)
                payload[i] = (char)(32 + (rng() % 95));
        }

        int num_packets = (int)(rng() % (int)payload.length()) + 1;
        if (num_packets < 1) num_packets = 1;

        int base_len = (int)payload.length() / num_packets;
        int rem = (int)payload.length() % num_packets;

        int offset = 0;
        for (int f = 0; f < num_packets; f++) {
            Packet pkt;
            pkt.flow_id = flow;
            pkt.fragment_id = f;
            pkt.num_fragments = num_packets;
            int this_len = base_len + (f < rem ? 1 : 0);
            pkt.length = this_len;
            pkt.data = payload.substr(offset, this_len);
            offset += this_len;
            g_data.packet_queue.push(pkt);
            total_packets++;
        }
    }

    (void)total_packets;

    printf("Percent attack  = %i\n", g_percent_attack);
    printf("Max data length = %i\n", g_max_data_length);
    printf("Num flow        = %i\n", g_num_flows);
    printf("Random seed     = %i\n", g_seed);
    printf("Num attack      = %i\n", g_data.total_attacks);
    fflush(stdout);

    auto start_time = std::chrono::high_resolution_clock::now();
    std::vector<std::thread> threads;
    for (int i = 0; i < g_num_threads; i++)
        threads.emplace_back(worker);
    for (auto& t : threads)
        t.join();
    auto end_time = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
                       end_time - start_time).count();

    uint64_t ops = g_total_ops.load();
    printf("Elapsed time = %f seconds\n", elapsed / 1000.0);
    printf("Num found = %lu\n", (unsigned long)ops);

    expli::TM<int>::exit();
    return 0;
}

File: benchmarks/rust/src/stamp/intruder.rs
use std::mem::size_of;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::Arc;
use tm::transaction;
use crate::Rng;
use super::Config;

const INTRUDER_MAX_DATA: usize = 256;
const INTRUDER_MAX_PACKETS: usize = 128;

// ── POD data structures (C layout) ──────────────────────────────────
#[derive(Clone)]
#[repr(C)]
struct Packet {
    flow_id: i64,
    fragment_id: i32,
    num_fragments: i32,
    length: i32,
    data: [i8; INTRUDER_MAX_DATA],
}

#[derive(Clone)]
#[repr(C)]
struct DecodedFlow {
    flow_id: i64,
    data_len: i32,
    data: [i8; INTRUDER_MAX_DATA * 2],
}

impl Default for Packet {
    fn default() -> Self {
        Packet {
            flow_id: -1,
            fragment_id: 0,
            num_fragments: 0,
            length: 0,
            data: [0i8; INTRUDER_MAX_DATA],
        }
    }
}

impl Default for DecodedFlow {
    fn default() -> Self {
        DecodedFlow {
            flow_id: -1,
            data_len: 0,
            data: [0i8; INTRUDER_MAX_DATA * 2],
        }
    }
}

// ── Shared TM-allocated state ──────────────────────────────────────
struct IntruderData {
    packet_queue: *mut u8,
    packet_q_head: *mut i32,
    packet_q_tail: *mut i32,
    #[allow(dead_code)]
    packet_q_capacity: i32,
    decoder_flows: *mut u8,
    fragment_storage: *mut i8,
    fragment_counts: *mut i32,
    decoded_queue: *mut u8,
    decoded_q_head: *mut i32,
    decoded_q_tail: *mut i32,
    decoded_q_capacity: i32,
    dictionary: Vec<String>,
    ops: &'static AtomicU64,
}

unsafe impl Send for IntruderData {}
unsafe impl Sync for IntruderData {}

// ── TX helpers ─────────────────────────────────────────────────────
fn get_packet(data: &IntruderData) -> Packet {
    transaction(|_tx| unsafe {
        let head = tm::tm_read_i32(data.packet_q_head);
        let tail = tm::tm_read_i32(data.packet_q_tail);
        if head < tail {
            let p = data.packet_queue.add(head as usize * size_of::<Packet>()) as *mut Packet;
            let pkt = Packet {
                flow_id: tm::tm_read_i64(std::ptr::addr_of_mut!((*p).flow_id)),
                fragment_id: tm::tm_read_i32(std::ptr::addr_of_mut!((*p).fragment_id)),
                num_fragments: tm::tm_read_i32(std::ptr::addr_of_mut!((*p).num_fragments)),
                length: tm::tm_read_i32(std::ptr::addr_of_mut!((*p).length)),
                data: {
                    let mut buf = [0i8; INTRUDER_MAX_DATA];
                    for i in 0..INTRUDER_MAX_DATA {
                        buf[i] = tm::tm_read_i8(std::ptr::addr_of_mut!((*p).data[i]));
                    }
                    buf
                },
            };
            tm::tm_write_i32(data.packet_q_head, head + 1);
            pkt
        } else {
            Packet::default()
        }
    })
}

fn process_decoder(data: &IntruderData, pkt: &Packet) {
    if pkt.flow_id < 0 { return; }
    transaction(|_tx| unsafe {
        let fid = pkt.flow_id;
        let idx = (fid - 1) as usize;
        let storage_off = (fid - 1) as usize * INTRUDER_MAX_PACKETS * INTRUDER_MAX_DATA
                          + pkt.fragment_id as usize * INTRUDER_MAX_DATA;

        for i in 0..(pkt.length as usize).min(INTRUDER_MAX_DATA) {
            tm::tm_write_i8(data.fragment_storage.add(storage_off + i), pkt.data[i]);
        }

        let prev = tm::tm_read_i32(data.fragment_counts.add(idx));
        tm::tm_write_i32(data.fragment_counts.add(idx), prev + 1);

        if prev + 1 == pkt.num_fragments {
            let df_ptr = data.decoder_flows.add(idx * size_of::<DecodedFlow>()) as *mut DecodedFlow;
            let mut total: usize = 0;
            for f in 0..pkt.num_fragments as usize {
                let off = (fid - 1) as usize * INTRUDER_MAX_PACKETS * INTRUDER_MAX_DATA
                         + f * INTRUDER_MAX_DATA;
                let flen = pkt.length as usize;
                for i in 0..flen.min(INTRUDER_MAX_DATA) {
                    if total >= INTRUDER_MAX_DATA * 2 { break; }
                    let byte = tm::tm_read_i8(data.fragment_storage.add(off + i));
                    tm::tm_write_i8(std::ptr::addr_of_mut!((*df_ptr).data[total]), byte);
                    total += 1;
                }
            }

            tm::tm_write_i32(std::ptr::addr_of_mut!((*df_ptr).data_len), total as i32);
            tm::tm_write_i64(std::ptr::addr_of_mut!((*df_ptr).flow_id), fid);

            let qtail = tm::tm_read_i32(data.decoded_q_tail);
            if qtail < data.decoded_q_capacity {
                let dst_ptr = data.decoded_queue.add(qtail as usize * size_of::<DecodedFlow>()) as *mut DecodedFlow;
                tm::tm_write_i64(std::ptr::addr_of_mut!((*dst_ptr).flow_id), fid);
                tm::tm_write_i32(std::ptr::addr_of_mut!((*dst_ptr).data_len), total as i32);
                for i in 0..total {
                    let byte = tm::tm_read_i8(std::ptr::addr_of_mut!((*df_ptr).data[i]));
                    tm::tm_write_i8(std::ptr::addr_of_mut!((*dst_ptr).data[i]), byte);
                }
                tm::tm_write_i32(data.decoded_q_tail, qtail + 1);
            }
            tm::tm_write_i32(data.fragment_counts.add(idx), 0);
        }
    });
}

fn get_complete(data: &IntruderData) -> DecodedFlow {
    transaction(|_tx| unsafe {
        let head = tm::tm_read_i32(data.decoded_q_head);
        let tail = tm::tm_read_i32(data.decoded_q_tail);
        if head < tail {
            let src = data.decoded_queue.add(head as usize * size_of::<DecodedFlow>()) as *mut DecodedFlow;
            let data_len = tm::tm_read_i32(std::ptr::addr_of_mut!((*src).data_len)) as usize;
            let df = DecodedFlow {
                flow_id: tm::tm_read_i64(std::ptr::addr_of_mut!((*src).flow_id)),
                data_len: data_len as i32,
                data: {
                    let mut buf = [0i8; INTRUDER_MAX_DATA * 2];
                    for i in 0..data_len.min(INTRUDER_MAX_DATA * 2) {
                        buf[i] = tm::tm_read_i8(std::ptr::addr_of_mut!((*src).data[i]));
                    }
                    buf
                },
            };
            tm::tm_write_i32(data.decoded_q_head, head + 1);
            df
        } else {
            DecodedFlow::default()
        }
    })
}

// ── Attack detection (non-TX) ──────────────────────────────────────
fn detect_attack(data: &[i8], data_len: usize, dictionary: &[String]) -> bool {
    let len = data_len.min(INTRUDER_MAX_DATA * 2);
    let mut lower = String::with_capacity(len);
    for &c in data.iter().take(len) {
        let bc = c as u8;
        if bc >= b'A' && bc <= b'Z' {
            lower.push((bc - b'A' + b'a') as char);
        } else {
            lower.push(bc as char);
        }
    }
    dictionary.iter().any(|word| lower.contains(word))
}

// ── Worker thread ──────────────────────────────────────────────────
fn worker(data: &IntruderData) {
    tm::tm_init_thread();
    loop {
        let pkt = get_packet(data);
        if pkt.flow_id >= 0 {
            process_decoder(data, &pkt);
        }
        let df = get_complete(data);
        if df.flow_id >= 0 {
            let _attack = detect_attack(&df.data, df.data_len as usize, &data.dictionary);
            data.ops.fetch_add(1, Ordering::Relaxed);
        } else if pkt.flow_id < 0 {
            break;
        }
    }
    tm::tm_exit_thread();
}

// ── Public entry point ─────────────────────────────────────────────
pub fn test() -> i32 {
    let mut fails = 0;
    // Test detect_attack with dictionary
    let dict: Vec<String> = vec![
        "attack".to_string(), "virus".to_string(), "malware".to_string(),
    ];
    let mut data = [0i8; 256];
    let msg = b"this is an attack message";
    for (i, &c) in msg.iter().enumerate() { data[i] = c as i8; }
    if !detect_attack(&data, msg.len(), &dict) {
        eprintln!("FAIL: 'attack' not detected"); fails += 1;
    }
    let safe = b"hello world this is fine";
    for (i, &c) in safe.iter().enumerate() { data[i] = c as i8; }
    if detect_attack(&data, safe.len(), &dict) {
        eprintln!("FAIL: false positive for safe message"); fails += 1;
    }
    // Test case-insensitive matching
    let mixed = b"Virus found here";
    for (i, &c) in mixed.iter().enumerate() { data[i] = c as i8; }
    if !detect_attack(&data, mixed.len(), &dict) {
        eprintln!("FAIL: case-insensitive 'Virus' not detected"); fails += 1;
    }
    if fails > 0 { eprintln!("intruder: {} test(s) failed", fails); }
    fails
}

pub fn run(config: &Config, _stop: &AtomicBool, ops: &AtomicU64) {
    let max_flows = config.num_packets;
    let max_packets_per_flow = config.max_length.min(INTRUDER_MAX_PACKETS);

    // ── Build dictionary ──
    let wordlist: &[&str] = &[
        "about", "attack", "back", "root", "system", "access",
        "all", "after", "also", "and", "any", "are", "but",
        "can", "come", "could", "did", "do", "each", "find",
        "first", "for", "from", "get", "go", "has", "have",
        "her", "here", "him", "his", "how", "into", "its",
        "just", "know", "like", "look", "make", "man", "may",
        "more", "most", "must", "new", "no", "not", "now",
        "old", "one", "only", "other", "our", "out", "over",
        "own", "part", "people", "said", "say", "see", "she",
        "shell", "should", "site", "some", "such", "take",
        "than", "that", "their", "them", "then", "there",
    ];
    let dictionary: Vec<String> = wordlist.iter().map(|s| s.to_string()).collect();

    // ── Allocate TM-safe flat arrays ──
    let packet_q_capacity = (max_flows * max_packets_per_flow + 1024) as i32;
    let packet_queue = addrspace::tm_region_calloc(packet_q_capacity as usize * size_of::<Packet>(), 1);
    let packet_q_head = addrspace::tm_region_calloc(size_of::<i32>(), 1) as *mut i32;
    let packet_q_tail = addrspace::tm_region_calloc(size_of::<i32>(), 1) as *mut i32;

    let decoder_flows = addrspace::tm_region_calloc(max_flows * size_of::<DecodedFlow>(), 1);
    let fragment_storage = addrspace::tm_region_calloc(max_flows * INTRUDER_MAX_PACKETS * INTRUDER_MAX_DATA, 1) as *mut i8;
    let fragment_counts = addrspace::tm_region_calloc(max_flows * size_of::<i32>(), 1) as *mut i32;

    let decoded_q_capacity = max_flows as i32;
    let decoded_queue = addrspace::tm_region_calloc(decoded_q_capacity as usize * size_of::<DecodedFlow>(), 1);
    let decoded_q_head = addrspace::tm_region_calloc(size_of::<i32>(), 1) as *mut i32;
    let decoded_q_tail = addrspace::tm_region_calloc(size_of::<i32>(), 1) as *mut i32;

    // ── Generate packets (single-threaded, no TM) ──
    let queue_ptr = packet_queue as *mut Packet;
    let mut total_packets: usize = 0;
    unsafe {
        let tail = &mut *packet_q_tail;
        let mut rng = Rng::new(config.seed);

        for flow in 1..=max_flows {
            let is_attack = (rng.next() % 100) < config.percent_attack as u64;

            let (payload, payload_len): ([i8; INTRUDER_MAX_DATA * 2], usize) = if is_attack {
                let sig_idx = (rng.next() % dictionary.len() as u64) as usize;
                let sig = dictionary[sig_idx].as_bytes();
                let slen = sig.len().min(INTRUDER_MAX_DATA * 2);
                let mut buf = [0i8; INTRUDER_MAX_DATA * 2];
                for i in 0..slen { buf[i] = sig[i] as i8; }
                (buf, slen)
            } else {
                let plen = ((rng.next() % config.max_length as u64) + 1) as usize;
                let plen = plen.min(INTRUDER_MAX_DATA * 2);
                let mut buf = [0i8; INTRUDER_MAX_DATA * 2];
                for i in 0..plen { buf[i] = (32 + (rng.next() % 95)) as i8; }
                (buf, plen)
            };

            let num_frags = ((rng.next() % payload_len as u64) + 1) as usize;
            let num_frags = num_frags.max(1).min(INTRUDER_MAX_PACKETS);

            let base_len = payload_len / num_frags;
            let rem = payload_len % num_frags;
            let mut offset = 0;

            for f in 0..num_frags {
                let pkt = &mut *queue_ptr.add(*tail as usize + total_packets + f);
                pkt.flow_id = flow as i64;
                pkt.fragment_id = f as i32;
                pkt.num_fragments = num_frags as i32;
                let this_len = base_len + if f < rem { 1 } else { 0 };
                pkt.length = this_len as i32;
                for i in 0..this_len.min(INTRUDER_MAX_DATA) {
                    pkt.data[i] = payload[offset + i];
                }
                offset += this_len;
            }
            total_packets += num_frags;
        }
        *tail += total_packets as i32;
    }

    println!("Total packets generated = {}", total_packets);
    println!("Flows = {}", max_flows);
    println!("Max length = {}", config.max_length);
    println!("Attack % = {}%", config.percent_attack);
    println!("Seed = {}", config.seed);
    println!();

    // ── Run workers ──
    let data = Arc::new(IntruderData {
        packet_queue,
        packet_q_head,
        packet_q_tail,
        packet_q_capacity,
        decoder_flows,
        fragment_storage,
        fragment_counts,
        decoded_queue,
        decoded_q_head,
        decoded_q_tail,
        decoded_q_capacity,
        dictionary,
        ops: unsafe { &*(ops as *const AtomicU64) },
    });

    let start = std::time::Instant::now();
    std::thread::scope(|s| {
        for _ in 0..config.threads {
            let d = Arc::clone(&data);
            s.spawn(move || worker(&d));
        }
    });
    let elapsed = start.elapsed().as_millis();

    let total_ops = ops.load(Ordering::Relaxed);
    println!("Elapsed time = {}.{:03} seconds", elapsed / 1000, elapsed % 1000);
    println!("Num found = {}", total_ops);

    // ── Cleanup ──
    addrspace::tm_region_free(packet_queue);
    addrspace::tm_region_free(packet_q_head as *mut u8);
    addrspace::tm_region_free(packet_q_tail as *mut u8);
    addrspace::tm_region_free(decoder_flows);
    addrspace::tm_region_free(fragment_storage as *mut u8);
    addrspace::tm_region_free(fragment_counts as *mut u8);
    addrspace::tm_region_free(decoded_queue);
    addrspace::tm_region_free(decoded_q_head as *mut u8);
    addrspace::tm_region_free(decoded_q_tail as *mut u8);
}
